core/src/main/scala/chisel3/dontTouch.scala

```diff
@@ -5,7 +5,8 @@ package chisel3
 import chisel3.experimental.{ChiselAnnotation, annotate, requireIsHardware}
 import firrtl.transforms.DontTouchAnnotation
 
-/** Marks that a signal should not be removed by Chisel and Firrtl optimization passes
+/** Marks that a signal is an optimization barrier to Chisel and the FIRRTL compiler. This has the effect of
```

Comment:
Is it obvious what "optimization barrier" means?
```suggestion
/** Marks that a signal is an optimization barrier to Chisel and the FIRRTL compiler.
  * The logical function on either side of the signal will be maintained. This has the effect of
```

Reply:
I was wondering the same thing.

Reply:
That would be helpful. I'm not 100% how to describe this, though, and I'm not sure if the suggestion captures it, either. (Since I'd say that optimizing across it would still preserve the logical function on either side to still be the same.) 

What about moving the `optimization barrier` language into a note and saying that this will both preserve the signal and prevent constants from being propagate through it?

Reply:
> prevent constants from being propagate through it?

Why would we want to guarantee that? I thought the whole point of `dontTouch` is to keep the signal around so that you can `bind` to it and see it in a waveform viewer. Why should a user care if we constant prop through the signal?
To me that would seem to restrictive for a Chisel-level API.

Reply:
> Why would we want to guarantee that?

:joy: [Because that's what this API is](https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/ConstantPropagation.scala#L579). :joy: I had the same reaction as you. Designers have wanted this and will use it to control downstream compilation. I agree that it's kind of a weird API and the name makes it more confusing.

Ideally, we'd have an API that gives you fine grained control over the product of: (1) signal is preserved and (2) optimization barrier.

This PR is narrowly about documenting exactly what the existing API does.

Reply:
> Why would we want to guarantee that?

Because that's what designers have asked for over time.

> the name makes it more confusing.

The name echos the same feature in synthesis tools ([1], [2], and see user guides for things like Design Compiler).

I absolutely agree that we need to distinguish between value semantics of a signal vs. optimization barrier, but historically we've just had the one mechanism so it ended up doing both 🙃 

[1] https://www.micro-ip.com/STA/dictionary_512_17/set_dont_touch.html
[2] https://inst.eecs.berkeley.edu/~cs250/fa09/handouts/tut5-dc.pdf